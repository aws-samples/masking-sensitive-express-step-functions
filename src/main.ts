import { App, CfnOutput, Duration, RemovalPolicy, Stack, StackProps, Tags } from 'aws-cdk-lib';
import { DataIdentifier, DataProtectionPolicy, LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import { DefinitionBody, LogLevel, Pass, StateMachine, StateMachineType } from 'aws-cdk-lib/aws-stepfunctions';
import { Construct } from 'constructs';

export class MaskingSensitiveExpressStepFunctionStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const piiDetectedLogGroup = new LogGroup(this, 'PiiDetectedLogGroup', {
      removalPolicy: RemovalPolicy.DESTROY, // For development purpose only
      retention: RetentionDays.ONE_DAY, // For development purpose only
    });

    const maskedLogGroup = new LogGroup(this, 'MaskedLogGroup', {
      dataProtectionPolicy: new DataProtectionPolicy({
        logGroupAuditDestination: piiDetectedLogGroup,
        description: 'Masks PII for name, address and email address',
        identifiers: [DataIdentifier.ADDRESS, DataIdentifier.EMAILADDRESS, DataIdentifier.NAME],
      }),

      removalPolicy: RemovalPolicy.DESTROY, // For development purpose only
      retention: RetentionDays.ONE_DAY, // For development purpose only
    });

    const loggingAndMaskingWorkflow = new StateMachine(this, 'LoggingAndMaskingWorkflow', {
      definitionBody: DefinitionBody.fromChainable(new Pass(this, 'LogInput')),
      stateMachineName: 'LoggingAndMaskingWorkflow',

      stateMachineType: StateMachineType.EXPRESS,
      timeout: Duration.seconds(5),
      logs: {
        destination: maskedLogGroup,
        level: LogLevel.ALL,
        includeExecutionData: true,
      },
    });

    // Create cdk output for the state machine
    new CfnOutput(this, 'LoggingAndMaskingWorkflowArn', {
      value: loggingAndMaskingWorkflow.stateMachineArn,
    });
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

const stack = new MaskingSensitiveExpressStepFunctionStack(app, 'masking-sensitive-express-step-functions-dev', {
  env: devEnv,
});
Tags.of(stack).add('project', 'masking-sensitive-express-step-functions');

app.synth();
