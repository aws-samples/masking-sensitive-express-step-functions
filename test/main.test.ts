import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MaskingSensitiveExpressStepFunctionStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new MaskingSensitiveExpressStepFunctionStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});