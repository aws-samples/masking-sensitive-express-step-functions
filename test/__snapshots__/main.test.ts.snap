// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Outputs": {
    "LoggingAndMaskingWorkflowArn": {
      "Value": {
        "Ref": "LoggingAndMaskingWorkflow5912D138",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LoggingAndMaskingWorkflow5912D138": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "LoggingAndMaskingWorkflowRoleDefaultPolicy8F375916",
        "LoggingAndMaskingWorkflowRoleC6C8C8FD",
      ],
      "Properties": {
        "DefinitionString": "{"StartAt":"LogInput","States":{"LogInput":{"Type":"Pass","End":true}},"TimeoutSeconds":5}",
        "LoggingConfiguration": {
          "Destinations": [
            {
              "CloudWatchLogsLogGroup": {
                "LogGroupArn": {
                  "Fn::GetAtt": [
                    "MaskedLogGroup196A9CAC",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": true,
          "Level": "ALL",
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "LoggingAndMaskingWorkflowRoleC6C8C8FD",
            "Arn",
          ],
        },
        "StateMachineName": "LoggingAndMaskingWorkflow",
        "StateMachineType": "EXPRESS",
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "LoggingAndMaskingWorkflowRoleC6C8C8FD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "LoggingAndMaskingWorkflowRoleDefaultPolicy8F375916": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LoggingAndMaskingWorkflowRoleDefaultPolicy8F375916",
        "Roles": [
          {
            "Ref": "LoggingAndMaskingWorkflowRoleC6C8C8FD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MaskedLogGroup196A9CAC": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DataProtectionPolicy": {
          "configuration": {
            "customDataIdentifier": [],
          },
          "description": "Masks PII for name, address and email address",
          "name": "data-protection-policy-cdk",
          "statement": [
            {
              "dataIdentifier": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dataprotection::aws:data-identifier/Address",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dataprotection::aws:data-identifier/EmailAddress",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dataprotection::aws:data-identifier/Name",
                    ],
                  ],
                },
              ],
              "operation": {
                "audit": {
                  "findingsDestination": {
                    "cloudWatchLogs": {
                      "logGroup": {
                        "Ref": "PiiDetectedLogGroup3F5A247B",
                      },
                    },
                  },
                },
              },
              "sid": "audit-statement-cdk",
            },
            {
              "dataIdentifier": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dataprotection::aws:data-identifier/Address",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dataprotection::aws:data-identifier/EmailAddress",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dataprotection::aws:data-identifier/Name",
                    ],
                  ],
                },
              ],
              "operation": {
                "deidentify": {
                  "maskConfig": {},
                },
              },
              "sid": "redact-statement-cdk",
            },
          ],
          "version": "2021-06-01",
        },
        "RetentionInDays": 1,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "PiiDetectedLogGroup3F5A247B": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "RetentionInDays": 1,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
